scipy/special/xsf/log_exp.h

```diff
@@ -65,4 +66,162 @@ T log_expit(T x) {
     return -std::log1p(std::exp(-x));
 };
 
+// Compute pow(1+x,y) accurately for real x and y.
+template <typename T>
+T pow1p_impl(T x, T y) {
+    auto out = stderr;
+    fprintf(out, "[pow1p] ========\n");
+    fprintf(out, "[pow1p] x=%.16e y=%.16e\n", x, y);
+
+    // The special values follow the spec of the pow() function defined in
+    // IEEE-754, Section 9.2.1.  The order of the `if` statements matters.
+    if (y == T(0)) { // pow(x, +/-0)
+        return T(1);
+    }
+    if (x == T(-1)) { // pow(+/-0, y)
+        if (std::isfinite(y) && y < 0) {
+            // TODO: signal divideByZero exception
+        }
+        return std::pow(T(0), y);
+    }
+    if (x == T(-2) && std::isinf(y)) { // pow(-1, +/-inf)
+        return T(1);
+    }
+    if (x == T(0)) { // pow(+1, y)
+        return T(1);
+    }
+    if (y == std::numeric_limits<T>::infinity()) { // pow(x, +inf)
+        return (x < 0 && x > -2) ? T(0) : std::numeric_limits<T>::infinity();
+    }
+    if (y == -std::numeric_limits<T>::infinity()) { // pow(x, -inf)
+        return (x < 0 && x > -2) ? std::numeric_limits<T>::infinity() : T(0);
+    }
+    if (std::isinf(x)) { // pow(+/-inf, y)
+        return std::pow(x, y);
+    }
+
+    // Up to this point, (1+x) = {+/-0, +/-1, +/-inf} have been handled, and
+    // and y = {+/-0, +/-inf} have been handled.  Next, we handle `nan` and
+    // y = {+/-1}.
+    if (std::isnan(x) || std::isnan(y)) {
+        return std::numeric_limits<T>::quiet_NaN();
+    }
+    if (y == T(1)) {
+        return T(1) + x;
+    }
+    if (y == T(-1)) {
+        return T(1) / (T(1) + x); // guaranteed no overflow
+    }
+
+    // Handle (1+x) < 0
+    T sign = T(1);
+    if (x < -1) {
+        if (std::fmod(y, T(1)) != T(0)) { // y is not an integer
+            // TODO: signal invalid operation exception
+            return std::numeric_limits<T>::quiet_NaN();
+        }
+        constexpr T two_over_eps = T(2) / std::numeric_limits<T>::epsilon();
+        // For 1 < (-x) <= 2^p = 2/eps, (1+x) is exact;
+        // For (-x) > 2^(2p) = (2/eps)^2, (1+x)^y ~= x^y.
+        // In both cases, we may delegate to pow() without loss of accuracy.
+        if (x >= -two_over_eps || x < -two_over_eps*two_over_eps) {
+            return std::pow(1+x, y);
+        }
+        // Otherwise, we take the "slow path" as in the case (1+x) > 0.
+        sign = (std::fmod(y, T(2)) == T(0)) ? T(1) : T(-1);
+    }
+
+    // Now x, y are finite and not equal to 0 or +/-1.
+    // To compute (1+x)^y, write |1+x| == (s+t) where s is equal to |1+x|
+    // rounded toward 1, and t is the (exact) rounding error.
+    volatile T s, t;
```

Comment:
This is not correct. Volatile should be reserved for when the value could be accessed by external processes.

Reply:
Indeed. I'm trying `volatile` here just to check if it would pass the test. (I learned the "volatile trick" from [here](https://github.com/scipy/scipy/blob/149f9fd32a1bdf1010944706b840c06816ea6a2f/scipy/special/xsf/cephes/dd_real.h#L72).)